opslevel/datasource_opslevel_filters.go

```diff
@@ -26,10 +28,13 @@ func datasourceFilters() *schema.Resource {
 
 func datasourceFiltersRead(d *schema.ResourceData, client *opslevel.Client) error {
 	resp, err := client.ListFilters(nil)
-	result := resp.Nodes
 	if err != nil {
 		return err
 	}
+	if resp == nil {
+		return fmt.Errorf("unexpected: listing filters returned nil")
```

Comment:
is this truly a problem to halt the execution of terraform plan?  Couldn't we defensively code this such that is the request to ListFilters returned an empty array that we store into the state an empty array too?

Additionally is it ever the case where `resp` == nil and `err` != nil ?

Reply:
>Couldn't we defensively code this such that is the request to ListFilters returned an empty array that we store into the state an empty array too?

We could, but in that case I think it would make more sense for `ListFilters` to not return a pointer, but rather return a slice. I have thought about the need to add this change for a while, since it is of 0 value for client applications to receive `WhateverConnection` types since they're essentially only useful for graphql requests whereas the consumers of opslevel-go are only concerned about what the actual data is. That would be out of scope for this PR. 

Right now the behaviour of terraform is to just crash if it encounters `resp == nil`, so this check is still better than what we had before. In the case that `resp != nil` and `len(resp.Nodes) == 0` the behaviour is completely unchanged. 

>Additionally is it ever the case where resp == nil and err != nil ?

With the opslevel-go code as it is now, this will never happen AFAICT. But there is no guarantee as to what opslevel-go will return as far as the terraform provider knows. It is totally possible that in the future we will accidentally make a change that results in `resp == nil && err != nil` and that would cause the terraform client to panic. 

`nilaway` supports annotations that we can add to opslevel-go to prevent it from alerting on this, but that would also not prevent the terraform provider from crashing on an erroneous change which would reduce how defensive our code is.

Reply:
```since it is of 0 value for client applications to receive WhateverConnection types since they're essentially only useful for graphql requests```

is not true - List calls are recursive so the connection object is needed to properly perform pagination

Reply:
> `since it is of 0 value for client applications to receive WhateverConnection types since they're essentially only useful for graphql requests`
> 
> is not true - List calls are recursive so the connection object is needed to properly perform pagination

OK, after re-thinking this it might be useful in certain applications that consume opslevel-go to retain the original pagination information that is given by a `Connection`. For example a use case would be to search in the first N pages of results but stop after that. There is also the benefit of `WhateverConnection` objects holding things like `[]ServiceId` instead of a massive `[]Service`. However I don't think any of our existing tools (CLI/TF/K8S) are taking advantage of this. 

In that case my idea of:

> it would make more sense for ListFilters to not return a pointer, but rather return a slice.

would not be useful.

I still think that the `nil` check should be retained however. I agree that it is bothersome, but we should either

1. Retain it, since a pointer is returned at the end of the day
2. Refactor `ListWhatever()` functions to always return `(ThingConnection, error)` instead of `(*ThingConnection, error)`.

Reply:
Ok lets leave the nil check for now then.